### ISSUE TITLE
Model profiling without model analyzer

### ISSUE THREAD

**Issue by ApoorveK** [triton-inference-server/server#4986]
**Description**
After deploying our ML model in triton server, I wanted to do the stress testing of the model by sending parallel requests but I got only one response form the triton server.

**Triton Information**
22.08

I am using Triton container

**To Reproduce**
Please write a multi-processing code to send multiple parallel requests to get the inference.

Model framework- TensorFlow
Model type- Ensemble
Number of models- 3
Sequence of models- preprocessing, inference, post-processing

**Expected behaviour**
For 'n' parallel requests, I should receive 'n' responses from the server which I am not getting right now and I should be able to analyse the model's performance and response time without using Triton's Model Analyzer. (since Model-Analyzer doesn't support ensemble model for now).

**Comment by kthui** [comment]
Hi @ApoorveK, can you share more details on how to replicate the issue?

**Comment by ApoorveK** [comment]
Hello @kthui thanks for replying. So I have the following Async client code, to send inference requests.
[ensembleModelInference.txt](https://github.com/triton-inference-server/server/files/9807235/ensembleModelInference.txt)

And this code is async client code, which take string as input and send inference requests for hosted ensemble model on triton inference server (this ensemble model comprise of total 3 models in which 2 are python backend models and 1 is of ONNX model for inference). For python backend models, we used default draft, which is given in doc [here](https://github.com/triton-inference-server/python_backend/blob/main/examples/preprocessing/model.py)

Now this code actually works for serial requests (giving 20 ms as time taken for inference) but now we have to test if model can handle concurrent requests. So would like your help/suggestions in making this code handling parallel requests easily. Also some following doubts:
* For two python models having similar name for class and same methods (initialise and execute) causes issue in performance.
* Do we have to make changes only in client code side or changes to be included in python models side as well in model directory, to make them handle parallel requests.
* For parallel inference request's stress testing for ensemble models, what should be procedure as currently Model-analyzer doesn't support ensemble models { see this [issue](https://github.com/triton-inference-server/model_analyzer/issues/400) }

**Comment by kthui** [comment]
Hi @ApoorveK, for async infer with the asyncio client, you can take a look at [this example](https://github.com/triton-inference-server/client/blob/main/src/python/examples/simple_grpc_aio_infer_client.py#L124) on how to do it. Both server and python_backend support parallel requests, so from what I can see, your client code will need some changes.

I am not familiar with python_backend related performance concern, @Tabrizian can you provide some context for point 1?

For point 3, I think it is better to ask on the [Model Analyzer's issue page](https://github.com/triton-inference-server/model_analyzer/issues), as they can provide more insight into the procedure.

**Comment by Tabrizian** [comment]
@ApoorveK

> For two python models having similar name for class and same methods (initialise and execute) causes issue in performance.

Can you elaborate more on what is the performance issue that you are observing?

> Do we have to make changes only in client code side or changes to be included in python models side as well in model directory, to make them handle parallel requests.

If you just want to send multiple requests to the model for profiling purposes, you can use [perf_analyzer](https://github.com/triton-inference-server/server/blob/main/docs/user_guide/perf_analyzer.md) with concurrencies of larger than 1.

**Comment by ApoorveK** [comment]
@Tabrizian thank you for responding, actually the issue was with some incorrect implementation of client code for triton model analyzer, as were trying to send input to triton server through post request, can you state the correct format for that!

we can do perf analyzer for the ensemble mode, but actually was looking to use model-analyzer for it's extra features.

**Comment by jbkyang-nvi** [comment]
Closing this issue due to lack of activity. If this issue needs follow-up, please let us know and we can reopen it for you.

**Comment by ApoorveK** [comment]
@jbkyang-nvi  @Tabrizian actually there is an issue with the current perf analyzer, while using it for ensemble models. As even if with no additional parameters and only input file, the perf analyzer is showing very high latency (10000-15000ms locally) as compared to brute force method (i.e. calculating time taken by server to serve request, by python client code locally which is around 40-60 ms)

**Comment by jbkyang-nvi** [comment]
@tgerdesnv @matthewkotila

**Comment by matthewkotila** [comment]
Hi @ApoorveK , sorry you're running into an issue using Perf Analyzer. We've created a ticket to investigate this.

But to help us with our investigation, could you provide more complete commands for reproducing your issue?

e.g. how to get the model downloaded from somewhere, put in a proper directory structure, launching the triton server, and the Perf Analyzer run command.

**Comment by ApoorveK** [comment]
@matthewkotila mostly i was running a similar example for ensemble model (which is being given is examples of triton inference server (https://github.com/triton-inference-server/python_backend/tree/main/examples/preprocessing). So while using perf analyzer (since model-analyzer doesn't support ensemble models for now) for it, the results were very weird, as the latency for the model interview is too high (as i mentioned above).

**Comment by matthewkotila** [comment]
Hi @ApoorveK can you be more specific with the exact commands you ran to download/generate the model repository you've referenced above? How did you get the `ensembleModel_bfsi`, `inferenceModel_bfsi`, `postProcessModel_bfsi`, and `preProcessModel_bfsi` models?

And can you also show the exact output that seemed incorrect?

And can you also give specific commands for how you ran the `brute force method (i.e. calculating time taken by server to serve request` mentioned above?

**Comment by ApoorveK** [comment]
Hello @matthewkotila Thanks for replying. So these 3 models  (inferenceModel_bfsi, postProcessModel_bfsi, and preProcessModel_bfsi) are actually specific use-case models which our organisation is using to try and test deployment on triton inference server. input for the 'postProcessModel_bfsi' is a string sentence, which being processes and passed to 'inferenceModel_bfsi' which give dictionary as output and then it get pre-processed through 'postProcessModel_bfsi' and gives final output after inference.

So with the model-analyzer container being started, with stressTestData.json containing input data as string.

Also, what exactly i did (the commands i ran) is being pasted along with the output actually. So the only help here i need is to set the current way for the supplied json file for model-analyzer and any other potential issues.

**Comment by matthewkotila** [comment]
Hi @ApoorveK, thanks for bringing up the potential Perf Analyzer issue. After thorough investigation, I've determined that this is expected behavior.

The reason you saw latency so high was because it includes the amount of time requests were waiting in the triton server queue. Why were there so many requests waiting in the queue? Because you ran Perf Analyzer with a concurrency of 400. That means roughly 399 requests of your model were all waiting in a queue while one was being run through the model at a time. That means latency will be quite high. Imagine the individual request latency for the 400th sent request. It was sent and entered the queue roughly the same time as the first 399 (since Perf Analyzer used 400 concurrent threads to send their first requests), but had to wait for the first 399 to be sent through the model, one at a time. This means the end-to-end latency for that 400th request will be quite high. This is essentially why the average latency is so high.

Let me know if you have any follow-up questions. Closing this issue.
